File: BaseSDK/Windows/Samples/Base/NYX650/DeviceSetFrameRate/DeviceSetFrameRate.cpp
#include <thread>
#include <iostream>
#include <thread>
#include "Scepter_api.h"

using namespace std;

time_t GetTimeStampMS();

int main()
{
	cout << "---DeviceSetFrameRate---"<< endl;

	//about dev
	uint32_t deviceCount;
	ScDeviceInfo* pDeviceListInfo = NULL;
	ScDeviceHandle deviceHandle = 0;
	ScStatus status = SC_OTHERS;

	//SDK Initialize
	status = scInitialize();
	if (status != ScStatus::SC_OK)
	{
		cout << "scInitialize failed status:" << status << endl;
		system("pause");
		return -1;
	}

	//1.Search and notice the count of devices.
	//2.get infomation of the devices. 
	//3.open devices accroding to the info.
	status = scGetDeviceCount(&deviceCount, 3000);
	if (status != ScStatus::SC_OK)
	{
		cout << "scGetDeviceCount failed! make sure pointer valid or called scInitialize()" << endl;
		system("pause");
		return -1;
	}
	cout << "Get device count: " << deviceCount << endl;
	if (0 == deviceCount)
	{
		cout << "scGetDeviceCount scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples."<< endl;
		system("pause");
		return -1;
	}

	pDeviceListInfo = new ScDeviceInfo[deviceCount];

	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		if (SC_CONNECTABLE != pDeviceListInfo[0].status)
		{
			cout << "connect status: " << pDeviceListInfo[0].status << endl;
			cout << "The device state does not support connection."<< endl;
			return -1;
		}
	}
	else
	{
		cout << "GetDeviceListInfo failed status:" << status << endl;
		return -1;
	}

	status = scOpenDeviceBySN(pDeviceListInfo[0].serialNumber, &deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "OpenDevice failed status:" << status << endl;
		return -1;
	}
    cout << "scOpenDeviceBySN status :" << status << endl;

	int frameRate = 5;
	status = scSetFrameRate(deviceHandle, frameRate);
	if (status != ScStatus::SC_OK)
	{
		cout << "scSetFrameRate failed status:" << status << endl;
		return -1;
	}
	cout << "set frame rate :" << frameRate <<" is OK."<< endl;

	//Starts capturing the image stream
	status = scStartStream(deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scStartStream failed status:" << status << endl;
		return -1;
	}

    cout << "Start testing the average frame rate for 30 seconds, Please wait patiently" << endl;
	//statistical frame rate
	for (;;)
	{
		ScFrameReady FrameReady = {};
		status = scGetFrameReady(deviceHandle, 1200, &FrameReady);
		if (status != ScStatus::SC_OK)
		{ 
			cout << "scGetFrameReady failed status:" << status << endl;
			continue;
		}

		if (1 == FrameReady.depth) 
		{
			ScFrame depthFrame = {};
			status = scGetFrame(deviceHandle, SC_DEPTH_FRAME, &depthFrame);
			if (depthFrame.pFrameData != NULL)
			{
                const int TESTPERIOD = 30;//30 senconds
				static int index = 0;
                static time_t start = GetTimeStampMS();

				time_t diff = GetTimeStampMS() - start;
                index++;
                if (diff > (TESTPERIOD * 1000))
                {
                    float fps = (index * TESTPERIOD * 1000.0f / diff) / TESTPERIOD;
                    index = 0;
					cout << fps << endl;
					break;
                }
			}
		}

	}

	status = scStopStream(deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scStopStream failed status:" << status << endl;
		return -1;
	}
	//1.close device
	//2.SDK shutdown
	status = scCloseDevice(&deviceHandle);
	if (status != ScStatus::SC_OK)
	{
		cout << "scCloseDevice failed status:" << status << endl;
		return -1;
	}
	status = scShutdown();
	if (status != ScStatus::SC_OK)
	{
		cout << "scShutdown failed status:" << status << endl;
		return -1;
	}
	cout << "---end---"<< endl;

	return 0;
}

time_t GetTimeStampMS()
{
    std::chrono::time_point<std::chrono::system_clock, std::chrono::milliseconds> tp = std::chrono::time_point_cast<std::chrono::milliseconds>(std::chrono::system_clock::now());
    time_t timestamp = tp.time_since_epoch().count();
    return timestamp;
}

File: MultilanguageSDK/Python/Samples/DS86/DeviceSetFrameRate/DeviceSetFrameRate.py
from pickle import FALSE, TRUE
import sys
currentPath =  sys.path[0]
pos = currentPath.find('Samples')
libpath = currentPath[:pos]
sys.path.append(libpath) #absolutely path

from API.ScepterDS_api import *
import time

camera = ScepterTofCam()

camera_count = camera.scGetDeviceCount(3000)
print("Get device count:", camera_count)
if camera_count <= 0:
    print("scGetDeviceCount scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples.")
    exit()

device_info=ScDeviceInfo()

if camera_count > 0:
    ret,device_infolist=camera.scGetDeviceInfoList(camera_count)
    if ret==0:
        device_info = device_infolist[0]
    else:
        print(' failed:' + ret)  
        exit()  
else: 
    print("there are no camera found")
    exit()

if  ScConnectStatus.SC_CONNECTABLE.value != device_info.status:
	print("connect status:",device_info.status)  
	print("Call scOpenDeviceBySN with connect status :",ScConnectStatus.SC_CONNECTABLE.value)
	exit()
else:
    print("serialNumber: "+str(device_info.serialNumber))
    print("ip: "+str(device_info.ip))
    print("connectStatus: "+str(device_info.status))

ret = camera.scOpenDeviceBySN(device_info.serialNumber)
if  ret != 0:
    print('scOpenDeviceBySN failed: ' + str(ret))
    exit()
	
print("scOpenDeviceBySN,status :"+ str(ret))

ret = camera.scSetFrameRate(5)
if ret == 0:
    print("Set frame rate:5")
else:
    print("scSetFrameRate failed status:"+str(ret))
    exit()

ret = camera.scSetWorkMode(ScWorkMode.SC_ACTIVE_MODE)
if ret == 0:
    print("scSetWorkMode SC_ACTIVE_MODE")
else:
    print("scSetWorkMode failed status:"+str(ret))
    exit()

ret = camera.scStartStream()
if  ret != 0:
    print("scStartStream failed:"+ str(ret))
    exit()

TESTPERIOD = 30
index = 0
start = 0
diff = 0
while 1:
    ret, frameready = camera.scGetFrameReady(c_uint16(1200))
    if ret != 0:
        print("scGetFrameReady failed status:"+ str(ret))
        continue

    if frameready.depth:
        ret, frame = camera.scGetFrame(ScFrameType.SC_DEPTH_FRAME)
        if ret == 0:
            if start == 0:
                start = int(time.time()*1000)
            diff = int(time.time()*1000) - start
            index += 1
            if diff > TESTPERIOD * 1000:
                fps = index / (diff / 1000)
                print("Frame rate: ", fps)
                break

ret = camera.scStopStream()
if  ret == 0:
    print("stopstream success")
else:
    print("stopstream failed", ret)

ret = camera.scCloseDevice()     
if  ret == 0:
    print("scCloseDevice successful")
else:
    print('scCloseDevice failed: ' + str(ret)) 
           